named_entity_extractor.py

```diff
@@ -0,0 +1,112 @@
+            {'name': 'President', 'type': 'PERSON', 'salience': 0.24969187378883362, 'mentions': [{'text': {'content': 'President', 'beginOffset': 24}, 'type': 'COMMON'}]},
+            {'name': 'Tim Curry', 'type': 'PERSON', 'metadata': {'wikipedia_url': 'https://en.wikipedia.org/wiki/Tim_Curry', 'mid': '/m/07rzf'}, 'salience': 0.1645415723323822, 'mentions': [{'text': {'content': 'Tim Curry', 'beginOffset': 10}, 'type': 'PROPER'}]},
+            {'name': '#Election2020', 'type': 'OTHER', 'salience': 0.04056445136666298, 'mentions': [{'text': {'content': '#Election2020', 'beginOffset': 35}, 'type': 'PROPER'}]}], 'language': 'en'}
+        
+        logging.error(
+            '_load_all_results is not implemented! returning dummy if cluster_id == 1234')
+        if cluster_id==1234:
+            return dummy_response
+        return None
+
+    def _generate_entity_list(self, ner_analysis_result):
+        return [entity['name'] for entity in ner_analysis_result['entities']]
+
+    def _analyze_entities(self, text_content):
+        """
+        Analyze Entities in a string using the GCP Language API
+
+        Structure of the returned values can be seen here:
+        https://cloud.google.com/natural-language/docs/basics
+
+        Args:
+        text_content The text content to analyze
+        """
+        logging.debug('making API call')
+
+        # Available types: PLAIN_TEXT, HTML
+        type_ = enums.Document.Type.PLAIN_TEXT
+
+        # Optional. If not specified, the language is automatically detected.
+        # For list of supported languages:
+        # https://cloud.google.com/natural-language/docs/languages
+        language = "en"
+        document = {"content": text_content,
+                    "type": type_, "language": language}
+
+        # Available values: NONE, UTF8, UTF16, UTF32
+        encoding_type = enums.EncodingType.UTF8
+
+        response = self.client.analyze_entities(
+            document, encoding_type=encoding_type)
+        return MessageToDict(response)
+
+    def get_cluster_text(self, cluster_id):
+        """ Get a Canonical Text result for a given cluster_id """
+        # TODO: Implement this
+        placeholder_text = 'Kermit vs John McCain for President. #StrangerThanFiction'
+        logging.error(
+            "get_cluster_text is not implemented! Returning placeholder text: '%s'", placeholder_text)
+        return placeholder_text
+
+    def get_entity_list_for_clusters(self, cluster_ids):
+        """ For ad clusters, get an entity: [cluster_id] map for further analysis"""
+        entity_cluster_map = defaultdict(list)
+        for cluster_id in cluster_ids:
+            cluster_text = self.get_cluster_text(cluster_id)
+
+            # Always try to fetch a result from storage if possible.
+            ner_analysis_result = self._load_all_results(
+                cluster_id) or self._analyze_entities(cluster_text)
```

Comment:
nit, please do a more verbose if branch on the result of self._load_all_results(). otherwise the API call could be easy to overlook

Reply:
Done.